### PR TITLE
fix(sink): force sink shuffle with the sink pk

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
force the same primary key be written into the same sink shard to make sure the sink pk mismatch compaction effective
https://github.com/risingwavelabs/risingwave/blob/6d88344c286f250ea8a7e7ef6b9d74dea838269e/src/stream/src/executor/sink.rs#L169-L198

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
Can you elaborate more in the PR body?

File: src/frontend/src/optimizer/plan_node/stream_sink.rs [hunk 1]
```diff
@@ -169,7 +169,19 @@ impl StreamSink {
                     }
                     _ => {
                         assert_matches!(user_distributed_by, RequiredDist::Any);
-                        RequiredDist::shard_by_key(input.schema().len(), input.expect_stream_key())
+                        if downstream_pk.is_empty() {
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/stream_sink.rs:172]: Is the sink compaction still needed if enforced shuffle?

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:172]: Yes, it happens on the shuffled data on the downstream fragment

**Review by StrikeW — APPROVED** [review]
LGTM

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**